okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java

```diff
@@ -134,4 +135,119 @@ private static String limiterKey(Request request) {
+     * {@link com.netflix.concurrency.limits.limiter.BlockingLimiter} to be asynchronous, maintaining a queue of
+     * currently waiting requests.
+     * <p>
+     * Upon a request finishing, we check if there are any waiting requests, and if there are we attempt to trigger some
+     * more.
+     */
+    final class ConcurrencyLimiter {
+        @GuardedBy("this")
+        private final Queue<SettableFuture<Limiter.Listener>> waitingRequests = new ArrayDeque<>();
+        @GuardedBy("this")
+        private Limiter<Void> limiter;
+        @GuardedBy("this")
+        private ScheduledFuture<?> timeoutCleanup;
+        private final String limiterKey;
+        private final Supplier<Limiter<Void>> limiterFactory;
+
+        ConcurrencyLimiter(String limiterKey, Supplier<Limiter<Void>> limiterFactory) {
+            this.limiterKey = limiterKey;
+            this.limiterFactory = limiterFactory;
+            this.limiter = limiterFactory.get();
+        }
+
+        synchronized ListenableFuture<Limiter.Listener> acquire() {
+            SettableFuture<Limiter.Listener> future = SettableFuture.create();
+            addSlowAcquireMarker(future);
+            waitingRequests.add(future);
+            processQueue();
+            return future;
+        }
+
+        synchronized void processQueue() {
+            while (!waitingRequests.isEmpty()) {
+                Optional<Limiter.Listener> maybeAcquired = limiter.acquire(NO_CONTEXT);
+                if (!maybeAcquired.isPresent()) {
+                    if (!timeoutScheduled()) {
+                        timeoutCleanup = scheduledExecutorService.schedule(
+                                this::resetLimiter, timeout.toMillis(), TimeUnit.MILLISECONDS);
+                    }
+                    return;
+                }
+                Limiter.Listener acquired = maybeAcquired.get();
+                SettableFuture<Limiter.Listener> head = waitingRequests.remove();
+                head.set(wrap(acquired));
+            }
+
+            if (timeoutScheduled()) {
+                timeoutCleanup.cancel(true);
+            }
+        }
+
+        private synchronized boolean timeoutScheduled() {
+            return timeoutCleanup != null && !timeoutCleanup.isDone() && !timeoutCleanup.isCancelled();
+        }
+
+        private synchronized void resetLimiter() {
+            log.warn("Timed out waiting to get permits for concurrency. In most cases this would indicate some kind of "
+                            + "deadlock. We expect that either this is caused by not closing response bodies "
+                            + "(there should be OkHttp log lines indicating this), or service overloading.",
+                    SafeArg.of("serviceClass", serviceClass),
+                    SafeArg.of("limiterKey", limiterKey),
+                    SafeArg.of("timeout", timeout));
+            leakSuspected.mark();
```

Comment:
can we tag this with the above args as well? (at least serviceClass and limiterKey)

Reply:
I have a suspicion that there's a reason why it wasn't tagged with all the things before. @j-baker might remember